CIPs/0016.md

```diff
@@ -0,0 +1,89 @@
+    
+    target                  = E164_PHONE_NUMBER / registry_id / celo_address
+    registry_id             = "StableToken" / "GoldToken" / "LockedGold" / ...
+    celo_address            = ( "0x" 40*40HEXDIG )
+    
+    endpoint                = token / function
+    token                   = "cUSD" / "CELO" / ...
+    function                = function_name "(" [ function_args ] ")"
+    function_name           = STRING
+    function_args           = function_arg *( "," function_arg )
+    function_arg            = ( NUMBER / STRING ) [ ":" TYPE ]
+
+    parameters              = parameter *( "&" parameter )
+    parameter               = key "=" value
+    key                     = "value" / "gas" / "gasPrice" / "feeCurrency" / "gatewayFee" / "gatewayFeeRecipient" / "localAmount" / "localCurrencyCode" / ...
+    value                   = NUMBER / STRING / celo_address
+
+### Semantics
+
+`TYPE` is a standard ABI type name, as defined in [Ethereum Contract ABI specification](https://solidity.readthedocs.io/en/develop/abi-spec.html). `STRING` is a URL-encoded unicode string of arbitrary length, where delimiters and the percentage symbol (`%`) are mandatorily hex-encoded with a `%` prefix.
+
+`E164_PHONE_NUMBER` is a standard E164 universal phone number, as defined by the International Telecommunication Union in [E164 Number Specification](https://www.itu.int/rec/T-REC-E.164).
+
+`prefix` is used for differentiating between user and contract `target`s and actions. `pay` and `request` are for user targets and intended for use with phone numbers or addresses and the `token` endpoint. `call` & `send` are for contract targets and intended for use with the registry/addresses and the `function` endpoint. 
+
+`token` is provided as an endpoint to more concisely encode celo protocol-aware ERC20 transfers that would otherwise be represented via a `function` endpoint.
+
+`registry_id` is provided to simplify and shorten celo protocol-aware contract interactions (to be eventually translated by the application to a registry-specified 40 byte address). 
+
+`function_args` have an optional `TYPE` because, in some cases, the application may have knowledge of the ABIs deeming the solidity types unnecessary. These can most likely be omitted when used with a celo registry contract via a `registry_id` target.
+
+`localAmount` is intended for use with `localCurrencyCode` and specifies a value in a locale-specific denomination (to be eventually translated by the application to a celo-specific denomination). Resolving `value` with `localAmount` is discretionary.
+
+###  Rationale
+
+Various QR schemes are used across implementations such as Valora, ContractKit, and likely others.
+
+Expanding the `target` spec to be inclusive of phone numbers and contract registry identifiers makes statically generated QR codes less likely to expire than if an address (which can be changed or compromised) is used. 
```

Comment:
This seems dangerous. What's the risk of a phone number not being mapped uniquely or the user not having enough quota to request a pepper vs. the risk of an address being expired?

Reply:
I think the other thing here is that this requires internet access to resolve, which hampers any sort of offline mode a little bit. If you're in a basement with your friend for dinner and want to send a payment that gets queued for later, this makes it difficult. Or sending via mesh networking, etc

Reply:
A transaction needs to be resolved online either way so I don't exactly see how an additional online step makes this prohibitive
"queued for later" could just include the lookup no?
I do recognize that including phone numbers here comes with a unique set of problems but imo this is one of the unique value props of Celo and app devs should be able to tap into it
Maybe it shouldn't be as "first-class" as it is here